spock-core/src/main/java/org/spockframework/mock/runtime/MockMakerRegistry.java

```diff
@@ -0,0 +1,154 @@
+package org.spockframework.mock.runtime;
+
+import org.spockframework.mock.CannotCreateMockException;
+import org.spockframework.mock.IMockObject;
+import org.spockframework.mock.ISpockMockObject;
+import org.spockframework.util.InternalSpockError;
+import org.spockframework.util.Nullable;
+import org.spockframework.util.ThreadSafe;
+import spock.mock.MockMakerId;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+import static org.spockframework.mock.runtime.IMockMaker.IMockCreationSettings;
+import static org.spockframework.mock.runtime.IMockMaker.ITypeMockability;
+
+@ThreadSafe
+public final class MockMakerRegistry {
+
+  private static final String NL = "\n";
+
+  public static MockMakerRegistry INSTANCE = createFromServiceLoader();
+
+  private final Map<String, IMockMaker> makerMap = new HashMap<>();
+
+  private final List<IMockMaker> makerList;
+
+  private static MockMakerRegistry createFromServiceLoader() {
+    ServiceLoader<IMockMaker> mockMakers = ServiceLoader.load(IMockMaker.class);
+    String preferredMockMakerId = System.getProperty(IMockMaker.MOCK_MAKER_PROPERTY);
+    return new MockMakerRegistry(mockMakers, preferredMockMakerId);
+  }
+
+  MockMakerRegistry(Iterable<IMockMaker> mockMakers, @Nullable String preferredMockMakerIdParam) {
+    List<IMockMaker> makerListLoc = new ArrayList<>();
+    mockMakers.forEach(m -> {
+      makerListLoc.add(m);
+      String id = m.getId();
+      if (id == null || !id.endsWith("-mock-maker")) {
+        throw new IllegalStateException("The IMockMaker '" + m + "' has an invalid ID: " + id);
+      }
+      if (m.getPriority() < 0) {
+        throw new IllegalStateException("The IMockMaker '" + id + "' has an invalid priority: " + m.getPriority());
```

Comment:
This should probably explain that negative priorities are disallowed.
If you change the validation a bit, you could reserve `0` for the `dynamicProxy` and make the sorting easier, the same goes for the `preferredMockMaker` which could be assigned to `1`.

Reply:
Done.